sdk/go/common/workspace/project.go

```diff
@@ -226,6 +342,62 @@ func ValidateProject(raw interface{}) error {
 	return errs
 }
 
+func InferFullTypeName(typeName string, itemsType *ProjectConfigItemsType) string {
```

Comment:
Pulumi YAML has a bunch of [type display logic](https://github.com/pulumi/pulumi-yaml/blob/e05b215177e550b3751c8c901eac1d0e4d6cd6e7/pkg/pulumiyaml/diags/types.go#L13). We should use it, probably necessitating moving it from pulumi/pulumi-yaml to sdk/go/common/???. I want to make sure we don't display the same type in different ways.

Reply:
Looking at that display logic, it looks like Pulumi YAML prints array types as `List<T>` while here we have chosen to output `array<T>` instead. This is due to the fact that the type definition has `type: array` so it would look weird/unexpected that types would be printed as `List<T>`

Also the type display logic looks too generic for `schema.Type`, the config types here are a small subset of those so I would like to keep it simple as is

Reply:
Yeh I'm going to lean to duplicating this display logic for now. Similar to comment below but config types != schema types, once we understand this domain better it might be something to clean up and unify.

Reply:
My concern is that users of Pulumi YAML will write
```yaml
config:
  foo:
    type: int
```
try to incorrectly use `foo`:

```yaml
resources:
  takesAListOfInt:
    properties:
      foo: ${foo}
```

and get an error:

```
foo: Cannot assign 'int' to 'List<int>'
```

If they give an a list of integers to the stack, they will get a different error:

```
Error: type mismatch between stack and project: found `array<int>` but expected `int`
```

Here the user need to understand that `List<int>` is equivalent to `array<int>` from (according to the user's perspective) the same program. I don't think that is a great experience. I choose `List` because it was what YAML used, but I'm not opposed to changing it. I just think it should be the same for the CLI and Pulumi YAML.

Reply:
> Here the user need to understand that List<int> is equivalent to array<integer>

Same with any other language though right? C# would say `int[]`, Go would say `[]int`, python would say `List[int]`.

I fully expect us to have to iterate heavily on what we print here as we start exposing more complex object types for config, there may be types there that aren't actually usable by yaml so I don't think we want to necessarily tie to the same printing. What makes sense for config in general might be too verbose/complex for yaml if it's restricted.

Reply:
It is the same for every other language. I guess the difference is that we have the ability unify the Pulumi CLI and Pulumi YAML messaging. I don't think users will expect to need to context switch between the Pulumi CLI and Pulumi YAML. Pulumi YAML `configuration` accepts a heavily restricted subset of types but Pulumi YAML itself operates on the full `schema.Type`. I'm struggling to image a scenario where it would make sense to pass in a config value that it too complicated to be handed to a Pulumi resource (and thus Pulumi YAML).

This doesn't have to be a blocker. We can come back to error messages after the feature is more developed.

Reply:
> We can come back to error messages after the feature is more developed.

Yeh totally happy to revisit this! I think once we think we've got type coverage for what we want to support at config level (and we know that's definitely more that what this first PR supports) we should look into both how the error messages look and unifying the use of those types and error messages in yaml.